File: 0126-word-ladder-ii/0126-word-ladder-ii.cpp
class Solution {
	unordered_map<string, int> mp;
	vector<vector<string>> res;
	string b;
private:
	void dfs(string word, vector<string> &seq) {
		if (word == b) {
			reverse(seq.begin(), seq.end());
			res.push_back(seq);
			reverse(seq.begin(), seq.end()); // to prevent changing the original sequence otherwise wrong element will be popped back
			return;
		}


		int steps = mp[word];
		int size = word.size();
		for (int i = 0; i < size; i++) {
			char original = word[i];
			for (char c = 'a'; c <= 'z'; c++) {
				word[i] = c;
				if (mp.find(word) != mp.end() && mp[word] == steps - 1) {
					seq.push_back(word);
					dfs(word, seq);
					seq.pop_back();
				}
			}
			word[i] = original;
		}
	}
public:
	vector<vector<string>> findLadders(string beginWord, string endWord, vector<string>& wordList) {
		int n = wordList.size();
		unordered_set<string> st(wordList.begin(), wordList.end());
		queue<string> q;
		q.push(beginWord);
		mp[beginWord] = 1;
		st.erase(beginWord);
		int size = beginWord.size();
		b = beginWord;

		while (!q.empty()) {
			string word = q.front();
			q.pop();

			int steps = mp[word];

			if (word == endWord) break;

			for (int i = 0; i < size; i++) {
				char original = word[i];
				for (char c = 'a'; c <= 'z'; c++) {
					word[i] = c;
					if (st.count(word)) {
						q.push(word);
						st.erase(word);
						mp[word] = steps + 1;
					}
				}
				word[i] = original;
			}
		}

		if (mp.find(endWord) != mp.end()) {
			vector<string> seq;
			seq.push_back(endWord);
			dfs(endWord, seq);
		}

		return res;
	}
};

File: 0126-word-ladder-ii/0126-word-ladder-ii.java
class Solution {
    public List<List<String>> findLadders(String beginWord, String endWord, List<String> wordList) {
        Set<String> wordSet = new HashSet<>(wordList);
        if (!wordSet.contains(endWord)) {
            return new ArrayList<>(); // End word is not in the word list
        }

        // Step 1: Use BFS to build a graph of all possible neighbors
        Map<String, List<String>> neighbors = new HashMap<>();
        wordSet.add(beginWord); // Ensure the begin word is part of the set
        for (String word : wordSet) {
            neighbors.put(word, new ArrayList<>());
        }
        for (String word : wordSet) {
            for (int i = 0; i < word.length(); i++) {
                char[] charArray = word.toCharArray();
                char original = charArray[i];
                for (char c = 'a'; c <= 'z'; c++) {
                    charArray[i] = c;
                    String newWord = new String(charArray);
                    if (wordSet.contains(newWord) && !newWord.equals(word)) {
                        neighbors.get(word).add(newWord);
                    }
                }
                charArray[i] = original; // Restore original character
            }
        }

        // Step 2: Bidirectional BFS
        Map<String, Integer> distance = new HashMap<>();
        Map<String, List<String>> paths = new HashMap<>();
        for (String word : wordSet) {
            paths.put(word, new ArrayList<>());
        }

        Queue<String> queue = new LinkedList<>();
        queue.offer(beginWord);
        distance.put(beginWord, 0);

        while (!queue.isEmpty()) {
            int size = queue.size();
            Set<String> currentLevelVisited = new HashSet<>();
            for (int i = 0; i < size; i++) {
                String word = queue.poll();
                int currentDist = distance.get(word);
                for (String neighbor : neighbors.get(word)) {
                    if (!distance.containsKey(neighbor)) {
                        distance.put(neighbor, currentDist + 1);
                        queue.offer(neighbor);
                        currentLevelVisited.add(neighbor);
                    }
                    if (distance.get(neighbor) == currentDist + 1) {
                        paths.get(neighbor).add(word);
                    }
                }
            }
            for (String visited : currentLevelVisited) {
                distance.put(visited, distance.get(visited));
            }
        }

        // Step 3: DFS to reconstruct all paths
        List<List<String>> results = new ArrayList<>();
        List<String> path = new ArrayList<>();
        path.add(endWord);
        dfs(endWord, beginWord, paths, path, results);

        return results;
    }

    private void dfs(String word, String beginWord, Map<String, List<String>> paths, List<String> path, List<List<String>> results) {
        if (word.equals(beginWord)) {
            List<String> temp = new ArrayList<>(path);
            Collections.reverse(temp); // Reverse to maintain the order
            results.add(temp);
            return;
        }
        for (String prevWord : paths.get(word)) {
            path.add(prevWord);
            dfs(prevWord, beginWord, paths, path, results);
            path.remove(path.size() - 1); // Backtrack
        }
    }
}
